In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\stefa\Documents\07-Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\stefa\\Documents\\07-Instacart Basket Analysis'

In [4]:
ords_prods_merge_new = pd.read_pickle(os.path.join(path, '02 data' , 'prepared data' , 'ords_prods_merge_new.pkl'))

In [5]:
df = ords_prods_merge_new

In [6]:
df.shape

(32434212, 18)

In [7]:
df.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0        1987   3139998      138            28                   6   
1        1989   1977647      138            30                   6   
2       11433    389851      709             2                   0   
3       12198    652770      764             1                   3   
4       12200   1813452      764             3                   4   

   orders_time_of_day (24 hours)  days_since_prior_order  add_to_cart_order  \
0                             11                     3.0                  5   
1                             17                    20.0                  1   
2                             21                     6.0                 20   
3                             13                     NaN                 10   
4                             17                     9.0                 11   

   reordered Check                 busiest days                 slowest days  \
0          0  both  Not one of the busiest days  Not one of the slowest days   
1          1  both  Not one of the busiest days  Not one of the slowest days   
2          0  both                 Busiest days  Not one of the slowest days   
3          0  both  Not one of the busiest days                 Slowest days   
4          1  both  Not one of the busiest days                 Slowest days   

  busiest_period_of_day  
0           Most orders  
1        Average orders  
2        Average orders  
3           Most orders  
4        Average orders

In [8]:
ords_prods_merge_new.drop('Unnamed: 0', axis = 1 )

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
32434207       49688      Fresh Foaming Cleanser        73             11   
32434208       49688      Fresh Foaming Cleanser        73             11   
32434209       49688      Fresh Foaming Cleanser        73             11   
32434210       49688      Fresh Foaming Cleanser        73             11   
32434211       49688      Fresh Foaming Cleanser        73             11   

          prices  order_id  user_id  order_number  orders_day_of_week  \
0            5.8   3139998      138            28                   6   
1            5.8   1977647      138            30                   6   
2            5.8    389851      709             2                   0   
3            5.8    652770      764             1                   3   
4            5.8   1813452      764             3                   4   
...          ...       ...      ...           ...                 ...   
32434207    13.5   1788356   200215             2                   0   
32434208    13.5   3401313   200377             1                   4   
32434209    13.5    809510   200873             5                   3   
32434210    13.5   2359893   200873             9                   3   
32434211    13.5   2385091   205926            11                   1   

          orders_time_of_day (24 hours)  days_since_prior_order  \
0                                    11                     3.0   
1                                    17                    20.0   
2                                    21                     6.0   
3                                    13                     NaN   
4                                    17                     9.0   
...                                 ...                     ...   
32434207                              9                     5.0   
32434208                             11                     NaN   
32434209                              8                    15.0   
32434210                             15                     5.0   
32434211                             15                     6.0   

          add_to_cart_order  reordered Check                 busiest days  \
0                         5          0  both  Not one of the busiest days   
1                         1          1  both  Not one of the busiest days   
2                        20          0  both                 Busiest days   
3                        10          0  both  Not one of the busiest days   
4                        11          1  both  Not one of the busiest days   
...                     ...        ...   ...                          ...   
32434207                 27          0  both                 Busiest days   
32434208                  5          0  both  Not one of the busiest days   
32434209                 12          0  both  Not one of the busiest days   
32434210                 11          1  both  Not one of the busiest days   
32434211                 35          0  both                 Busiest days   

                         slowest days busiest_period_of_day  
0         Not one of the slowest days           Most orders  
1         Not one of the slowest days        Average orders  
2         Not one of the slowest days        Average orders  
3                        Slowest days           Most orders  
4                        Slowest days        Average orders  
...                               ...                   ...  
32434207  Not one of the slowest days       

In [9]:
ords_prods_merge_new.groupby('department_id').order_number.mean()

department_id
1     15.457687
2     17.277920
3     17.179756
4     17.811403
5     15.213779
6     16.439806
7     17.225773
8     15.340520
9     15.895474
10    20.197148
11    16.170828
12    15.887622
13    16.583304
14    16.757377
15    16.165037
16    17.663250
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

When comparing the results for the entire dataframe differ from those of the subset, results from entire dataframe differ slightly from the subset. Most showed only a difference in the decimal place, but some were off by between 1 & 3 values. EX. A department listed as having a mean of 22.9 when looking at the entire dataframe would have a mean of 25.5 in the subset.

Step 5

In [10]:
ords_prods_merge_new['max_order'] = ords_prods_merge_new.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\stefa\AppData\Local\Temp\ipykernel_12384\3036065107.py:1: FutureWarning: The provided callable <function max at 0x0000026E67E35E40> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge_new['max_order'] = ords_prods_merge_new.groupby(['user_id'])['order_number'].transform(np.max)


Adding flags for loyal, new and regular customers.

In [11]:
ords_prods_merge_new.loc[ords_prods_merge_new['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
ords_prods_merge_new.loc[(ords_prods_merge_new['max_order'] <= 40) & (ords_prods_merge_new['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
ords_prods_merge_new.loc[ords_prods_merge_new['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
ords_prods_merge_new['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

In [15]:
ords_prods_merge_new.groupby('loyalty_flag')['prices'].mean()

loyalty_flag
Loyal customer      10.388747
New customer        13.294370
Regular customer    12.496203
Name: prices, dtype: float64

There is some noticable differences in prices. Loyal customers tend to purchase less expensive items then new and regular customers.

Step 6

In [16]:
ords_prods_merge_new['avg_spending_user'] = ords_prods_merge_new.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\stefa\AppData\Local\Temp\ipykernel_12384\1509482039.py:1: FutureWarning: The provided callable <function max at 0x0000026E67E35E40> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge_new['avg_spending_user'] = ords_prods_merge_new.groupby(['user_id'])['order_number'].transform(np.max)


Adding flags for high and low spenders

In [17]:
ords_prods_merge_new.loc[ords_prods_merge_new['avg_spending_user'] > 10, 'spending_flag'] = 'Low spender'

In [18]:
ords_prods_merge_new.loc[ords_prods_merge_new['avg_spending_user'] <= 10, 'spending_flag'] = 'High spender'

In [19]:
#Confirming results
ords_prods_merge_new[['user_id', 'avg_spending_user', 'spending_flag']].head(20)

user_id  avg_spending_user spending_flag
0       138                 32   Low spender
1       138                 32   Low spender
2       709                  5  High spender
3       764                  3  High spender
4       764                  3  High spender
5       777                 26   Low spender
6       825                  9  High spender
7       910                 12   Low spender
8      1052                 20   Low spender
9      1052                 20   Low spender
10     1379                 10  High spender
11     1480                 43   Low spender
12     1480                 43   Low spender
13     1480                 43   Low spender
14     1494                 13   Low spender
15     1494                 13   Low spender
16     1494                 13   Low spender
17     1540                 28   Low spender
18     1540                 28   Low spender
19     1540                 28   Low spender

Step 7

In [20]:
ords_prods_merge_new['ords_freq_median'] = ords_prods_merge_new.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\stefa\AppData\Local\Temp\ipykernel_12384\298894402.py:1: FutureWarning: The provided callable <function median at 0x0000026E67F654E0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge_new['ords_freq_median'] = ords_prods_merge_new.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


Adding flags for frequent, non-frequent and regular customers

In [21]:
ords_prods_merge_new.loc[ords_prods_merge_new['ords_freq_median'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [22]:
ords_prods_merge_new.loc[(ords_prods_merge_new['ords_freq_median'] <= 20) & (ords_prods_merge_new['ords_freq_median'] > 10), 'frequency_flag'] = 'Regular customer'

In [23]:
ords_prods_merge_new.loc[ords_prods_merge_new['ords_freq_median'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [24]:
#Confirming results
ords_prods_merge_new[['user_id', 'ords_freq_median', 'frequency_flag']].head(20)

user_id  ords_freq_median     frequency_flag
0       138               8.0  Frequent customer
1       138               8.0  Frequent customer
2       709               8.0  Frequent customer
3       764               9.0  Frequent customer
4       764               9.0  Frequent customer
5       777              11.0   Regular customer
6       825              20.0   Regular customer
7       910               6.0  Frequent customer
8      1052              10.0  Frequent customer
9      1052              10.0  Frequent customer
10     1379              20.0   Regular customer
11     1480               7.0  Frequent customer
12     1480               7.0  Frequent customer
13     1480               7.0  Frequent customer
14     1494               7.0  Frequent customer
15     1494               7.0  Frequent customer
16     1494               7.0  Frequent customer
17     1540               7.0  Frequent customer
18     1540               7.0  Frequent customer
19     1540               7.0  Frequent customer

In [26]:
ords_prods_merge_new.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_4.8.pkl'))